pyomo/gdp/plugins/partition_disjuncts.py

```diff
@@ -0,0 +1,850 @@
+        default=False,
+        domain=bool,
+        description="""Boolean indicating whether or not to transform so that
+        the transformed model will still be valid when fixed Vars are
+        unfixed.""",
+        doc="""
+        If True, the transformation will create a correct model even if fixed
+        variables are later unfixed. That is, bounds will be calculated based
+        on fixed variables' bounds, not their values. However, if fixed
+        variables will never be unfixed, a possibly tigher model will result,
+        and fixed variables need not have bounds.
+
+        Note that this has no effect on fixed BooleanVars, including the
+        indicator variables of Disjuncts. The transformation is always correct
+        whether or not these remain fixed.
+        """
+    ))
+    CONFIG.declare('compute_bounds_method', ConfigValue(
+        default=compute_fbbt_bounds,
+        description="""Function which takes an expression, a Block containing
+        the global constraints of the original problem, and a configured
+        solver, and returns both a lower and upper bound for the expression.""",
+        doc="""Callback for computing bounds on expressions, in order to bound
+        the auxilary variables created by the transformation. Some
+        pre-implemented options include
+            * compute_fbbt_bounds (the default), and
+            * compute_optimal_bounds
+        or you can write your own callback which accepts an Expression object,
+        a model containing the variables and global constraints of the original
+        instance, and a configured solver and returns a tuple (LB, UB) where
+        either element can be None if no valid bound could be found.
+        """
+    ))
+    CONFIG.declare('compute_bounds_solver', ConfigValue(
+        default=None,
+        description="""Solver object to pass to compute_bounds_method.
+        This is required if you are using 'compute_optimal_bounds'.""",
+        doc="""
+        Configured solver object for use in the compute_bounds_method.
+
+        In particular, if compute_bounds_method is 'compute_optimal_bounds',
+        this will be used to solve the subproblems, so needs to handle
+        non-convex problems if any Disjunctions contain nonlinear constraints.
+        """
+    ))
+    def __init__(self):
+        super(PartitionDisjuncts_Transformation, self).__init__()
+        self.handlers = {
+            Constraint:  self._transform_constraint,
+            Var:         False, # these will be already dealt with--we add
+                                # references to them before we call handlers.
+            BooleanVar:  False,
+            Connector:   False,
+            Expression:  False,
+            Suffix:      False,
+            Param:       False,
+            Set:         False,
+            SetOf:       False,
+            RangeSet:    False,
+            Disjunction: False, # this actually isn't possible in this
```

Comment:
An exception would be unreachable code because in this transformation we have to transform from the outside in, so when we encounter an active Disjunction, we transform it before we call the handler. So I think you're right, I'll omit this completely because if it happens, it's actually a bug with the implementation.